application_store/db/queries/application/queries.py

```diff
@@ -266,25 +269,96 @@ def submit_application(application_id) -> Applications:
             "jsonb_blob": ApplicationSchema().dump(application),
             "type_of_application": application_type,
         }
-        stmt = postgres_insert(AssessmentRecord).values([row])
 
-        upsert_rows_stmt = stmt.on_conflict_do_nothing(index_elements=[AssessmentRecord.application_id]).returning(
-            AssessmentRecord.application_id
+        existing_application = db.session.scalar(
+            select(AssessmentRecord)
+            .where(
+                AssessmentRecord.application_id == row["application_id"],
+                AssessmentRecord.is_withdrawn == False,  # noqa: E712
+            )
+            .options(load_only(AssessmentRecord.jsonb_blob))
         )
 
-        result = db.session.execute(upsert_rows_stmt)
-
-        # Check if the inserted application is in result
-        inserted_application_ids = [item.application_id for item in result]
-        if not len(inserted_application_ids):
-            current_app.logger.warning(
-                "Application already exists in the database: {app_id}", extra=dict(app_id=row["application_id"])
-            )
+        if existing_application and fund.funding_type == "UNCOMPETED":
+            # For uncompeted funds, the application may already exist and this may be a resubmission.
+            field_map = {}
+            changed_fields = set()
+            for form in existing_application.jsonb_blob["forms"]:
+                for section in form["questions"]:
+                    for field in section["fields"]:
+                        field_map[field["key"]] = field["answer"]
+
+            for form in row["jsonb_blob"]["forms"]:
+                for section in form["questions"]:
+                    for field in section["fields"]:
+                        if field["answer"] != field_map.get(field["key"]):
+                            changed_fields.add(field["key"])
+                            if "history_log" in field:
+                                field["history_log"].append(
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                )
+                            else:
+                                field["history_log"] = [
+                                    {datetime.now(tz=timezone.utc).isoformat(): field_map[field["key"]]}
+                                ]
+
+            if changed_fields:
+                row["workflow_status"] = WorkflowStatus.CHANGE_RECEIVED
```

Comment:
This one is up to you... I don't like `helpers` functions very much, but this update row logic might be nice if extracted so it can be tested individually.

Reply:
Yeah good shout - I've moved it out into a separate function and added some tests.